plugins/outputs/stackdriver/counter_cache.go

```diff
@@ -0,0 +1,96 @@
+package stackdriver
+
+import (
+	"path"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+
+	monpb "google.golang.org/genproto/googleapis/monitoring/v3"
+	tspb "google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type CounterCache struct {
+	sync.RWMutex
+	cache map[string]*CounterCacheEntry
+	log   telegraf.Logger
+}
+
+type CounterCacheEntry struct {
+	LastValue *monpb.TypedValue
+	StartTime *tspb.Timestamp
+}
+
+func (cce *CounterCacheEntry) Reset(ts *tspb.Timestamp) {
+	// always backdate a reset by -1ms, otherwise stackdriver's API will hate us
+	cce.StartTime = tspb.New(ts.AsTime().Add(time.Millisecond * -1))
+}
+
+func (cc *CounterCache) get(key string) (*CounterCacheEntry, bool) {
+	cc.RLock()
+	value, ok := cc.cache[key]
+	cc.RUnlock()
+	return value, ok
```

Comment:
You can simplify this to
```suggestion
	cc.RLock()
	defer cc.RUnlock()
	return cc.cache[key]
```
Using defer here is strongly recommended in general as it prevents returning with a lock held.

Reply:
👍

Reply:
actually this seems to end up needing to be:

```
func (cc *counterCache) get(key string) (*counterCacheEntry, bool) {
	cc.RLock()
	defer cc.RUnlock()
	value, ok := cc.cache[key]
	return value, ok
}
```

...because the compiler really does't like trying to directly return the map lookup even though it seems like it _should?_

```
plugins/outputs/stackdriver/counter_cache.go:35:2: not enough arguments to return
	have (*counterCacheEntry)
	want (*counterCacheEntry, bool)
```

Reply:
Ah right, I always forget about this stupidity. :-)